autoload/db.vim

```diff
@@ -528,3 +601,125 @@ function! db#clobber_dbext(...) abort
   endfor
   return opts
 endfunction
+
+function! s:vim_job.cb(job, data) dict abort
+  if type(a:data) ==? type(0)
+    let self.exit = 1
+    let self.exit_status = a:data
+    return self.call_cb_if_finished()
+  endif
+  let self.output .= a:data
+endfunction
+
+function! s:vim_job.close_cb(channel) dict abort
+  let self.close = 1
+  return self.call_cb_if_finished()
+endfunction
+
+function! s:vim_job.call_cb_if_finished() abort
+  if self.close && self.exit
+    return self.callback(split(self.output, "\n", 1), self.exit_status)
+  endif
+endfunction
+
+function! s:nvim_job_cb(jobid, data, event) dict abort
+  if a:event ==? 'exit'
+    return self.callback(self.output, a:data)
+  endif
+  call extend(self.output, a:data)
+endfunction
+
+function! s:job_run(cmd, callback, stdin) abort
+  if has('nvim')
+    let jobid = jobstart(a:cmd, {
+          \ 'on_stdout': function('s:nvim_job_cb'),
+          \ 'on_stderr': function('s:nvim_job_cb'),
+          \ 'on_exit': function('s:nvim_job_cb'),
+          \ 'output': [],
+          \ 'callback': a:callback,
+          \ 'stdout_buffered': 1,
+          \ 'stderr_buffered': 1,
+          \ })
+
+    if !empty(a:stdin)
+      let stdin = a:stdin
+      if filereadable(stdin)
+        let stdin = readfile(stdin, 'b')
+      endif
+      call chansend(jobid, stdin)
+      call chanclose(jobid, 'stdin')
+    endif
+
+    return jobid
+  endif
+
+  if exists('*job_start')
+    let fn = copy(s:vim_job)
+    let fn.callback = a:callback
+    let opts = {
+          \ 'out_cb': fn.cb,
+          \ 'err_cb': fn.cb,
+          \ 'exit_cb': fn.cb,
+          \ 'close_cb': fn.close_cb,
+          \ 'mode': 'raw'
+          \ }
+
+    if has('patch-8.1.350')
+      let opts['noblock'] = 1
+    endif
+
+    let job = job_start(a:cmd, opts)
+
+    if !empty(a:stdin)
+      let stdin = a:stdin
+      if filereadable(stdin)
+        let stdin = exists('*readblob') ? readblob(stdin) : readfile(stdin, 'B')
+      endif
```

Comment:
@kristijanhusak as read this, if the file doesn't exist it will silently be treated as a query instead? And if the query happens to match a filename it will will silently use the file contents instead? This is a completely unnecessary compromise.

Furthermore I checked and neither `readblob()` nor the `B` flag of `readfile()` are supported on Vim 8.0.0000. If I can't find a quick fix tomorrow, I'm going to roll this back until I have time to sort it out.

(See also #116, another job bug which got me to dig in to begin with.)

Reply:
> as read this, if the file doesn't exist it will silently be treated as a query instead?

Yeah, but I think stdin is always a filename (temp one), if I didn't miss anything. If that's the case, `stdin` variable name is misleading, it should be named differently, and we should avoid setting the default. Do you know if we have situation where this function gets stdin as non-filepath?

> And if the query happens to match a filename it will will silently use the file contents instead?

Yeah this would be the situation, but I think this situation is no longer applicable. Could be a leftover.

> Furthermore I checked and neither readblob() nor the B flag of readfile() are supported on Vim 8.0.0000. If I can't find a quick fix tomorrow, I'm going to roll this back until I have time to sort it out.

I don't remember if I tested 8.0.0000, but I did test 8.1.* , and maybe 8.0.* , but I don't remember which version.
Since 9.0 is the latest and 8.2 is stable for some time now (at least on some channels), is support for 8.0.* necessary?

Reply:
> > as read this, if the file doesn't exist it will silently be treated as a query instead?
> 
> Yeah, but I think stdin is always a filename (temp one), if I didn't miss anything. If that's the case, `stdin` variable name is misleading, it should be named differently, and we should avoid setting the default. Do you know if we have situation where this function gets stdin as non-filepath?
> 
> > And if the query happens to match a filename it will will silently use the file contents instead?
> 
> Yeah this would be the situation, but I think this situation is no longer applicable. Could be a leftover.

`db#systemlist()` appears to pass a query string directly.

> > Furthermore I checked and neither readblob() nor the B flag of readfile() are supported on Vim 8.0.0000. If I can't find a quick fix tomorrow, I'm going to roll this back until I have time to sort it out.
> 
> I don't remember if I tested 8.0.0000, but I did test 8.1.* , and maybe 8.0.* , but I don't remember which version. Since 9.0 is the latest and 8.2 is stable for some time now (at least on some channels), is support for 8.0.* necessary?

As you may recall, I decided to drop 7.x support after careful deliberation. I don't want to drop 8.0 on a whim, at least not if all it buys us is saving an `else`.

Reply:
> db#systemlist() appears to pass a query string directly.

Yeah, but I guess only when called directly, not through a `DB` command.

> As you may recall, I decided to drop 7.x support after careful deliberation. I don't want to drop 8.0 on a whim, at least not if all it buys us is saving an else.

Yeah, because of `jobs`. Is there a way to read binary file on 8.0 ?

Reply:
We don't need to, use `'in_io': 'file'`.

Reply:
Yeah that will work. Neovim still needs to read the file, but it has the binary `readfile`.
Do you want to tackle this or should I make a PR?

Reply:
Make a PR. I reverted to fix #116 so you'll need to start by reverting that reversion. Expect a small conflict around `s:systemlist()`.